### PR TITLE
Unify product routing

### LINKED ISSUE #185: Unify product routing
Right now we have two styles of endpoints for products. For example, for product details following routes apply:
 * products/{code}
 * products-by-slug/{slug}

In order to unify them, my proposal is as follows:
 * products/by-code/{code}
 * products/by-slug/{slug}

Rest of endpoints would work similar way, so instead of `taxon-products-by-slug/` we would have `taxon-products/by-slug/`

### PR DESCRIPTION
**Pull request by mamazu**
Closes #185

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1,3 +1,17 @@
+# UPGRADE FROM 1.0.0-beta.22 to 1.0.0-beta.23
```
Comment (GSadee) [inline UPGRADE.md]: For now, there is no `beta.22`

File: UPGRADE.md [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+|:--------------------------------------|:-------------------------------------|
+| `products/{code}`                     | `products/by-code/{code}`            |
+| `products-by-slug/{slug}`             | `products/by-slug/{slug}`            |
+| `products/{code}/reviews`             | `products/by-code/{code}/reviews`    |
+| `products-reviews-by-slug/{slug}`     | `products/by-slug/{slug}/reviews`    |
+| `taxon-products/{code}`               | `taxon-products/by-code/{taxonCode}` |
+| `taxon-products-by-slug/{taxonSlug}`  | `taxon-products/by-slug/{taxonSlug}` |
+
+
+
```
Comment (GSadee) [inline UPGRADE.md]: Unnecessary blank lines

File: src/Resources/config/routing/productBySlug.yml [hunk 3]
```diff
@@ -1,31 +1,64 @@
...
     path: /product-reviews-by-slug/{slug}
     methods: [POST]
     defaults:
         _controller: sylius.shop_api_plugin.controller.product.add_review_by_slug_action
     requirements:
         slug: .+
 
-sylius_shop_api_product_show_catalog_by_slug:
+sylius_shop_api_product_show_catalog_by_slug_deprecated:
     path: /taxon-products-by-slug/{taxonSlug}
```
Comment (GSadee) [inline src/Resources/config/routing/productBySlug.yml]: `/taxon-products-by-slug/{slug}`?
to be consistent with `/taxon-products/by-code/{code}`

File: UPGRADE.md (hunk 1 above)
Comment (mamazu) [reply UPGRADE.md]: What version is it?

File: src/Resources/config/routing/productBySlug.yml [hunk 4]
```diff
@@ -1,31 +1,64 @@
...
+sylius_shop_api_product_show_details_by_slug:
+    path: /products/by-slug/{slug}
+    methods: [GET]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.product.show_product_details_by_slug_action
+    requirements:
+        slug: .+
+
+sylius_shop_api_product_show_catalog_by_slug:
+    path: /taxon-products/by-slug/{taxonSlug}
```
Comment (GSadee) [inline src/Resources/config/routing/productBySlug.yml]: This is a place where I should have aded this comment:
`/taxon-products/by-slug/{slug}`? to be consistent with `/taxon-products/by-code/{code}`

File: UPGRADE.md (hunk 1 above)
Comment (GSadee) [reply UPGRADE.md]: Now it is `beta.21`

File: src/Resources/config/routing/productBySlug.yml (hunk 4 above)
Comment (mamazu) [reply src/Resources/config/routing/productBySlug.yml]: Correct, yeah. For that I fixed the documentation. I couldn't get it to work with the different name. It was like this before.

File: src/Resources/config/routing/productBySlug.yml [hunk 5]
```diff
@@ -1,31 +1,64 @@
...
+    methods: [GET]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.product.show_product_catalog_by_taxon_slug_action
+    requirements:
+        taxonSlug: .+
+
+sylius_shop_api_product_show_reviews_by_slug:
+    path: /product/by-slug/{slug}/reviews
```
Comment (mamazu) [inline src/Resources/config/routing/productBySlug.yml]: This is going to cause a problem. See the related issue https://github.com/Sylius/Sylius/issues/10293#issuecomment-480795973

File: src/Resources/config/routing/productByCode.yml [hunk 6]
```diff
@@ -1,23 +1,48 @@
...
+
+# New routes
+sylius_shop_api_product_show_details_by_code:
+    path: /products/by-code/{code}
+    methods: [GET]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.product.show_product_details_by_code_action
+
+sylius_shop_api_product_show_catalog_by_code:
+    path: /taxon-products/by-code/{code}
```
Comment (GSadee) [inline src/Resources/config/routing/productByCode.yml]: `code` -> `taxonCode`?
Now, it is not consistent with routing by slug, and also with you change in `UPGRADE` file

Comment (mamazu) [reply src/Resources/config/routing/productByCode.yml]: I will change that.

Comment (GSadee) [reply src/Resources/config/routing/productByCode.yml]: But now, there are `taxon-products/by-code/{code}` and `taxon-products/by-slug/{taxonSlug}` routings, so they are still not consistent. Sorry for being annoying, it is of course not very important 😃

Comment (mamazu) [reply src/Resources/config/routing/productByCode.yml]: I'll try to get it done during the day.

File: src/Resources/config/routing/productByCode.yml [hunk 7]
```diff
@@ -1,23 +1,48 @@
-sylius_shop_api_product_show_details_by_code:
+sylius_shop_api_product_show_details_by_code_deprecated:
```
Comment (lchrusciel) [inline src/Resources/config/routing/productByCode.yml]: as we are allowed to do bc-breaks, you have documented change and updated docs, I think we can just remove this bc layer. It is not even tested right now and such an upgrade should be relatively easy

**Comment by lchrusciel** [comment]
Thank you, @mamazu! :1st_place_medal:

**Pull request merged**